florestad/src/json_rpc/blockchain.rs

```diff
@@ -44,6 +46,72 @@ impl<Blockchain: RpcChain> RpcImpl<Blockchain> {
             .get_block(&blockhash)
             .map_err(|_| JsonRpcError::BlockNotFound)
     }
+
+    /// Retrieves the height of the block that was mined in the given timestamp.
+    ///
+    /// `timestamp` has an alias, 0 will directly refer to 1231006505, the time of the genesis block, returning 0.
+    pub async fn get_block_height_by_timestamp(
+        &self,
+        timestamp: u32,
+        confidence: &RescanConfidence,
+    ) -> Result<u32, JsonRpcError> {
+        /// Simple helper to avoid code reuse.
+        fn get_block_time<BlockChain: RpcChain>(
+            provider: &RpcImpl<BlockChain>,
+            at: u32,
+        ) -> Result<u32, JsonRpcError> {
+            let hash = provider.get_block_hash(at)?;
+            let block = provider.get_block_header(hash)?;
+            Ok(block.time)
+        }
+
+        let genesis_timestamp = genesis_block(self.network).header.time;
+
+        if timestamp == 0 || timestamp == genesis_timestamp {
+            return Ok(0);
+        };
+
+        let (tip_height, _) = self
+            .chain
+            .get_best_block()
+            .map_err(|_| JsonRpcError::BlockNotFound)?;
+
+        let tip_time = get_block_time(self, tip_height)?;
+
+        if timestamp < genesis_timestamp || timestamp > tip_time {
+            return Err(JsonRpcError::InvalidTimestamp);
+        }
+
+        let adjusted_target = timestamp.saturating_sub(confidence.val());
+
+        let mut high = tip_height;
+        let mut low = 0;
+        let max_iters = tip_height.ilog2() + 1;
+        for _ in 0..max_iters {
+            let cut = (high + low) / 2;
+
+            let block_timestamp = get_block_time(self, cut)?;
+
+            if block_timestamp == adjusted_target {
+                debug!("Find by timestamp found a precise block; returning {cut}");
```

Comment:
`Find by timestamp` is meant to refer to this function? I think you don't need this, since the actual log will have the function name in there.